### ISSUE TITLE
[RFC] Add an extensible TorchFT process-group registry

### ISSUE THREAD

**Issue by Skuld7451** [pytorch/torchtitan#4206]
# RFC 0001: Extensible TorchFT process-group registry

## Summary

Replace the hard-coded process-group `if/elif` chain in TorchTitan's TorchFT
manager with a public registry. TorchTitan keeps the existing Gloo, NCCL, and
MCCL behavior, while a device integration can register another backend before
constructing `TorchFTManager`.

## Problem

`TorchFTManager` previously owned every backend constructor. Supporting another
backend therefore required modifying or subclassing the manager, even though
backend construction is independent from training orchestration. That couples
optional runtime dependencies to the generic integration and makes every new
backend a core-code change.

### Public registration API

`torchtitan/experiments/torchft/process_group_registry.py:18-42` validates and
registers a factory under a normalized name:

```python
def register_process_group_factory(
    name: str,
    factory: ProcessGroupFactory,
) -> None:
    normalized_name = name.strip().lower()
    if not normalized_name:
        raise ValueError("Process group name must not be empty")
    if not callable(factory):
        raise TypeError("Process group factory must be callable")
    ...
    _PROCESS_GROUP_FACTORIES[normalized_name] = factory
```

The same callable may be registered again without effect. Replacing an
existing name with a different callable fails explicitly, so import order
cannot silently change the selected implementation.

`torchtitan/experiments/torchft/process_group_registry.py:45-60` provides
creation and introspection:

```python
def create_process_group(name: str, timeout: timedelta) -> Any:
    normalized_name = name.strip().lower()
    factory = _PROCESS_GROUP_FACTORIES.get(normalized_name)
    if factory is None:
        ...
        raise ValueError(...)
    return factory(timeout)


def registered_process_group_names() -> tuple[str, ...]:
    return tuple(sorted(_PROCESS_GROUP_FACTORIES))
```

These functions are exported from
`torchtitan/experiments/torchft/__init__.py:12-26`.

### Built-in backends preserve existing behavior

Gloo, NCCL, and MCCL remain lazily constructed at
`torchtitan/experiments/torchft/process_group_registry.py:63-92`. Optional
packages such as `torchcomms` are imported only when their backend is selected.

```python
register_process_group_factory("gloo", _create_gloo)
register_process_group_factory("nccl", _create_nccl)
register_process_group_factory("mccl", _create_mccl)
```

The manager delegates only the construction step at
`torchtitan/experiments/torchft/manager.py:95-96`:

```python
process_group_timeout = timedelta(milliseconds=config.process_group_timeout_ms)
pg = create_process_group(config.process_group, process_group_timeout)
```

### Extension usage

An integration registers its backend during its initialization path:

```python
from torchtitan.experiments.torchft import register_process_group_factory

register_process_group_factory("custom", create_custom_process_group)
```

The factory receives one `datetime.timedelta` positional argument and returns
a TorchFT-compatible process group. User-facing usage is documented in
`torchtitan/experiments/torchft/README.md:83-123`.

## Compatibility and scope

- Existing `gloo`, `nccl`, and `mccl` configuration values keep their prior
  constructors and timeout semantics.
- Backend imports stay lazy; installing one backend does not make all optional
  backend runtimes mandatory.
- This proposal does not implement a device process group. It only provides the
  extension point through which an integration supplies one.
- The change does not require the LocalSGD, synchronous integration, or healing
  changes.

## Validation

`torchtitan/experiments/torchft/tests/test_process_group_registry.py:21-97`
covers built-in Gloo/NCCL construction, MCCL argument preservation, external
registration and normalization, idempotency, conflicting registration, invalid
inputs, and diagnostics for unknown backends.

The registry module contains six focused tests. The same full TorchFT test
directory used by CI passed against this branch:

```text
pytest torchtitan/experiments/torchft/tests -v
7 passed
```

The seventh test is the pre-existing checkpoint test, which guards against
regression outside the newly added registry module.
